pkg/util/error_list.go

```diff
@@ -45,10 +45,67 @@ func (list errorList) Error() string {
 	return result
 }
 
+// Except returns any errors (or nested errors, if the list contains
+// nested []error slices) for which all fns return false. If no errors
+// remain a nil list is returned. The resulting list will have all
+// nested []error slices flattened as a side effect.
+func (list ErrorList) Except(fns ...ErrMatchFunc) ErrorList {
+	if len(list) == 0 {
+		return nil
+	}
+	lists := []ErrorList{list}
+	flat := ErrorList{}
+	for len(lists) > 0 {
+		list := lists[0]
+		lists = lists[1:]
+		for i := range list {
+			err := list[i]
+			if list, ok := err.(ErrorList); ok {
+				lists = append(lists, list)
+				continue
+			}
+			if !matchesError(err, fns...) {
+				flat = append(flat, err)
+			}
+		}
+	}
+	if len(flat) == 0 {
+		return nil
+	}
+	return flat
+}
+
+// matchesError returns true if any ErrMatchFunc returns true
+func matchesError(err error, fns ...ErrMatchFunc) bool {
+	for _, fn := range fns {
+		if fn(err) {
+			return true
+		}
+	}
+	return false
+}
+
 // SliceToError converts an []error into a "normal" error, or nil if the slice is empty.
 func SliceToError(errs []error) error {
 	if len(errs) == 0 {
 		return nil
 	}
-	return errorList(errs)
+	return ErrorList(errs)
+}
+
+// ErrMatchFunc is used to match errors.
+type ErrMatchFunc func(error) bool
+
+// FilterError returns the error (or, if err is a slice of errors, the subset
+// of that slice) if all provided match functions return false. This can be used
+// to eliminate known-good errors from a list (such as io.EOF or os.PathNotFound)
+// of errors to determine whether any true error occurred.
+func FilterError(err error, fns ...ErrMatchFunc) error {
```

Comment:
Do you really need to handle singleton errors in the same way?  This would be simpler if it were just operating on []error.